### PR TITLE
[Feature/1722] - Change CAS3 assessment search by using status directly in DB query

### PR DESCRIPTION
**Pull request by sudhakar-plr**
# Changes in this PR
Refere [1722](https://trello.com/c/TTDY9BDm/1722-improve-performance-of-loading-referrals-during-the-receive-journey) for more information.

This PR to overcome performance problem in searching CAS3 assessment. 

**Current**:  the search is happening by selecting all assessment and then filter by assessment status.

**Change**:  This PR is to add the additional WHERE clause to an in existing query using 'status'(assement-status). Additionally merging existing 2 CAS3 assessment search query into single query.

Added indexes to applications(crn), temporary_accommodation_applications(probation_region_id) to improve assessment search query performance.

The status where clause is derived by following the existing logic as below

```
**Status**                                    **DB Query**
rejected                                          assessment.decision = 'REJECTED'
closed                                            assessment.decision = 'ACCEPTED' AND applications.completed_at !=null
ready_to_place                             assessment.decision = 'ACCEPTED' AND applications.completed_at =null
in_review                                       assessment.decision = null AND assessments.allocated_to_user_id !=null
unallocated                                   assessment.decision = null AND assessments.allocated_to_user_id =null
```

We may need to make changes in front end to search assessment with status whenever the user change the status view page.

# Release checklist

As part of our continuous deployment strategy we must ensure that this work is
ready to be released at any point. Before merging to `main` we must first
confirm:

## Pre merge checklist

- [ ] Have all changes to any dependencies been deployed to both `preprod` and
    `production` in a backwards compatible way? (This includes our API,
    infrastructure, third-party integrations and any secrets or environment variables)
- [ ] Has a required data migration been included in this change or been done in
    advance?

## Post merge checklist

Once we've merged we now need to release this through to production before
considering it done.

[Find the build-and-deploy job in CircleCI](https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-temporary-accommodation-ui)
and work through the following steps:

- [ ] Has the product manager asked to provide approval for this feature?
  - [ ] Has the product manager approved?
- [ ] Manually approve release to preprod
- [ ] Verify change released to preprod
- [ ] Manually approve release to prod
- [ ] Verify change released to production

Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible.

### CODE REVIEW

**Review by tahb — APPROVED** [review]
Looks good to me. I double checked locally and it appears to be backwards compatible too 👍 

I've left two minor comments, we should probably improve the naming one before this goes in but I'll approve now for convenience.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/AssessmentTest.kt
Comment (sudhakar-plr) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/AssessmentTest.kt:513]: Since its a integration test, i would like to keep the set of record consistent for all tests so that the 'given' data is same for all test and specific expected results. In this way, am not manipulating test data but make sure the results are correct in all situation.

Its local integration test db and would like to keep the data as realistic as possible.

Anyway, i have removed where its not needed, but i would prefer the data setup is same for all testcases.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/AssessmentService.kt
Comment (sudhakar-plr) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/AssessmentService.kt]: done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/AssessmentEntity.kt
Comment (tahb) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/AssessmentEntity.kt:163]: [I noted that we don't think this query will fall foul of the same problems as booking search.](https://mojdt.slack.com/archives/C03K0HB0LBE/p1702386439174639?thread_ts=1702383217.041969&cid=C03K0HB0LBE)

[It was interesting to learn that CAS1 have taken a different approach here with a new status column](https://github.com/ministryofjustice/hmpps-approved-premises-api/pull/1119). [They appear to be tackling the problem in the same way as our other idea](https://github.com/ministryofjustice/hmpps-approved-premises-api/pull/1119/files#diff-f5add0cc3f1580e3f29dbb37dda68174daacd19376c2d5b86e2c8ffd7ca11c77R3) but as we're going in second we're actually now in the position of introducing diverging behaviour.

Given that work, what do you think is best for the codebase as a whole?

I wonder if it's:

1. merge this as is for short term performance improvements
2. migrate to using a status column to make our handling of status more explicit and consistent

Comment (sudhakar-plr) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/AssessmentEntity.kt:163]: We could make changes to create new status column to make our sql straight forward, but this will take time as we need to understand flow/journey of assessment and may take time to fix it.

I personally think, we can go with current work and then we can plan for adding the status later. I even thinking whether merge this code now or after christmas when we both wont be available if any issue.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/AssessmentTest.kt
Comment (tahb) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/AssessmentTest.kt:513]: That's reasonable for improved clarity between tests, I hasn't considered that.

**Comment by sudhakar-plr** [comment]
We are just holding this PR not to merge because we need to complete UI change to make sure of the performance improvement.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/AssessmentController.kt
Comment (davidatkinsuk) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/AssessmentController.kt:40]: can we remove filterByStatuses now?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/AssessmentController.kt
Comment (davidatkinsuk) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/AssessmentController.kt:77]: i think you can pass through a null perPage value and the default will be resolved in the PaginationUtils, as we did for CAS1 below (i.e. no need to set default here, and i think we can remove injection of pagination.default-page-size)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/AssessmentEntity.kt
Comment (davidatkinsuk) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/AssessmentEntity.kt:127]: nice to see the very similar queries merged into one 👍

**Review by davidatkinsuk — APPROVED** [review]
nice to see the CAS1 and CAS3 code converging again in the controller 👍 

a few minor comments from me to consider. approved from a code quality POV but can't really comment on CAS3 status resolution (although it looks logical to me!)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/AssessmentController.kt
Comment (davidatkinsuk) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/AssessmentController.kt:81]: can we remove these sort and filer methods now?

Comment (sudhakar-plr) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/AssessmentController.kt:81]: We(Cas3) dont store the person name in the table, so currently personName sort is using the existing sort functionality.

Comment (davidatkinsuk) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/AssessmentController.kt:81]: Ah yes, sorry i didn't spot you were still calling the sort method

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/AssessmentController.kt
Comment (sudhakar-plr) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/AssessmentController.kt]: done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/AssessmentController.kt
Comment (sudhakar-plr) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/AssessmentController.kt:81]: removed un-used methods

**Pull request merged**
